book/src/proposals/durable-tx-nonces.md

```diff
@@ -29,64 +38,75 @@ TODO: svgbob this into a flowchart
-                                which will replace this "parameter" */
-  if spend_hash == stored_hash
-    if !sysvar.recent_blockhashes.is_empty()
-      stored_hash = sysvar.recent_blockhashes[0]
-      success
-    else
-      error
-  else
-    error
+  elif state != Initialized
+    error BadState
+  if sysvar.recent_blockhashes.is_empty()
+    error EmptyRecentBlockhashes
+  stored_hash = sysvar.recent_blockhashes[0]
+  success
+WithdrawInstruction(to, lamports)
+  if state == Uninitialized
+    if !signers.contains(owner)
+      error MissingRequiredSignatures
+  elif state == Initialized
+    if lamports != account.balance && lamports + rent_exempt > account.balance
+      error InsufficientFunds
+  account.balance -= lamports
+  to.balance += lamports
+  success
 ```
 
 A client wishing to use this feature starts by creating a nonce account and
 depositing sufficient lamports as to make it rent-exempt. The resultant account
 will be in the `Uninitialized` state with no stored hash and thus unusable.
 
-To begin using the account an `Initialize` instruction is executed on it,
-advancing its state to `Initialized` and storing a durable nonce, chosen by the
-contract from the `recent_blockhashes` sysvar, in the data field.
-
-When making use of a durable nonce, the client must first query its value from
-account data. A transaction is now constructed in the normal way, but with the
-following additional requirements:
-
-  1) The durable nonce value is used in the `recent_blockhash` field
-  2) A `Nonce` instruction is issued (first?)
-  3) The appropriate transaction flag (TBD) is set, signaling that the usual
-hash age check should be skipped and the previous requirements enforced
+The `Nonce` instruction is used to request that a new nonce be stored for the
+calling account. The first `Nonce` instruction run on a newly created account
+will drive the account's state to `Initialized`. As such, a `Nonce` instruction
+MUST be issued before the account can be used.
 
-For the `Nonce` instruction to succeed:
-  1) The nonce account MUST have been advanced to the `Initialized` state,
-ensuring that a valid nonce has been stored.
-  2) The nonce value in the transaction's `recent_blockhash` field MUST match
-the value stored in the nonce account data field. In doing so, the client
-commits to the stored nonce value by signing the transaction.
-  3) The nonce value MUST NOT reside in the `recent_blockhashes` sysvar, thus
-preventing replay by deleting, then recreating the account.
-If these requirements are met, the contract replaces the stored nonce with a
-new one from the current values in the `recent_blockhashes` sysvar.
-
-To discard a nonce account, the client should include a `Nonce` instruction in
-a transaction which withdraws all lamports, leaving a zero balance and making
-it eligible for deletion.
+To discard a `NonceAccount`, the client should issue a `Withdraw` instruction
```

Comment:
Withdraw must wait until the current Nonce is no longer in recent_blockhashes...

Reply:
Yep.  Forgot to add that here. Will remedy

Reply:
https://github.com/solana-labs/solana/pull/6725/commits/8497e152aed83bb03ff601adacb9c28e0f892ce9